pennylane/clusters.py

```diff
@@ -0,0 +1,455 @@
+
+    >>> x = qml.init.strong_ent_layers_normal(3, 2)
+    >>> cost(x)
+    tensor(0.9092, dtype=torch.float64, grad_fn=<SumBackward0>)
+    """
+    if hasattr(x, "interface") and x.interface is not None:
+        if x.interface == "tf":
+            import tensorflow as tf
+
+            return apply(tf.reduce_sum, x)
+
+        if x.interface == "torch":
+            import torch
+
+            return apply(torch.sum, x)
+
+        if x.interface in ("autograd", "numpy"):
+            from autograd import numpy as np
+
+            return apply(np.sum, x)
+
+        raise ValueError("Unknown interface {}".format(x.interface))
+
+    import numpy as np
+
+    return apply(np.sum, x)
+
+
+def _get_dot_func(interface):
+    """Helper function for :func:`~.dot` to determine
+    the correct dot product function depending on the QNodeCluster
+    interface"""
+    if interface == "tf":
+        import tensorflow as tf
+
+        return lambda a, b: tf.tensordot(a, b, 1)
+
+    if interface == "torch":
+        import torch
+
+        return torch.matmul
+
+    if interface in ("autograd", "numpy"):
+        from autograd import numpy as np
+
+        return np.dot
+
+    if interface is None:
+        import numpy as np
+
+        return np.dot
+
+    raise ValueError("Unknown interface {}".format(interface))
+
+
+def dot(x, y):
+    r"""Lazily perform the dot product between arrays, tensors, and :class:`QNodeCluster`.
+
+    Using this function, lazy dot products can be computed between two :class:`QNodeCluster`
+    objects, or a :class:`QNodeCluster` object and an array/tensor object. In the latter
+    case, only one-dimensional arrays/tensors are supported.
+
+    Args:
+        x (array or tensor or QNodeCluster): A QNode cluster of independent QNodes,
+            or an array/tensor object.
+        y (array or tensor or QNodeCluster): A QNode cluster of independent QNodes,
+            or an array/tensor object.
+
+    .. seealso:: :func:`~.apply`, :func:`~.sum`
+
+    **Example:**
+
+    We can create a QNodeCluster using :func:`~.map`:
+
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> obs_list = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliZ(0) @ qml.PauliZ(1)]
+    >>> qnodes = qml.map(qml.templates.StronglyEntanglingLayers, obs_list, dev, interface="torch")
+
+    The returned QNodeCluster contains 2 QNodes, as we mapped the :func:`~.StronglyEntanglingLayers`
+    over a list of two observables:
+
+    >>> len(qnodes)
+    2
+
+    For the cost function, we now perform the dot product between a vector of coefficients
+    and the QNodeCluster:
+
+    >>> coeffs = torch.tensor([0.32, -0.2], dtype=torch.double)
+    >>> cost = qml.dot(coeffs, qnodes)
+
+    Note that ``cost`` is equivalent to computing :math:`\langle 0 | U(\theta)^\dagger H U(\theta) | 0\rangle`
+    where
+
+    * :math:`U(\theta)` is the unitary applied by the strongly entangling layers, and
+    * :math:`H = 0.32 X\otimes Z - 0.2 Z \otimes Z`.
```

Comment:
If a ``Notes`` section is created, this could be moved there (now this seems to constitute as theory and might be slightly lost inside of the example).